bindings/c/include/opendal.h

```diff
@@ -862,6 +928,10 @@ char *opendal_list_entry_name(const struct opendal_list_entry *self);
  */
 void opendal_list_entry_free(struct opendal_list_entry *ptr);
 
+intptr_t opendal_reader_read(const struct opendal_reader *self, uint8_t *buf, uintptr_t len);
```

Comment:
We should return error here, can be done after https://github.com/apache/incubator-opendal/pull/3250

Reply:
At first, it was returning "opendal_error", but upon further consideration, I think it would be better to return the size of the buffer being read. 🤔

Reply:
> I think it would be better to return the size of the buffer being read. 🤔

Yes, we should return the value of `written_size` upon success and an error message upon failure.

Reply:
You could have a `opendal_result_reader_read` with an `opendal_code` and a `size`.

After the error message pr is merged, we will change `opendal_code` to `opendal_error`

Reply:
> You could have a `opendal_result_reader_read` with an `opendal_code` and a `size`.

By the way, do we need to allocate on the success case after migrating to `opendal_error`?

Reply:
I don't think we need to allocate. Because on success, the error is a nullptr and all other values are on stack.